### PR TITLE
Array.__setitem__ doesn't update the chunksize, resulting in bad graph traversal

### LINKED ISSUE #3730: Dask array setitem with differently chunked data provides wrong shape
```python
>>> import dask.array as da
>>> x = da.zeros(5, chunks=2)
>>> m = da.zeros(5, chunks=3)
>>> x[m>0] = 1
>>> print x.shape, x.compute().shape
(5,) (4,)
```

This should probably unifiy chunks before slicing

### PR DESCRIPTION
**Pull request by itamarst**
Fixes #3730.

Mutation based on another array can result in chunk size changing. Chunk size is basis for deciding what keys to search the graph for. Therefore we need to make sure chunk size is up-to-date.

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -1337,6 +1337,10 @@ def __setitem__(self, key, value):
             self.dtype = y.dtype
             self.dask = y.dask
             self.name = y.name
+            # In order to figure out what keys to use, __dask_keys__ needs to
+            # know what chunks to expect. The chunks to expect are y.chunks, so
+            # we need to change our chunks.
```
Comment (mrocklin) [inline dask/array/core.py]: I recommend that we remove this comment

File: dask/array/tests/test_slicing.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 from operator import getitem
 
 import pytest
+from hypothesis import given, strategies as st, example, settings
```
Comment (mrocklin) [inline dask/array/tests/test_slicing.py]: This should be discussed in a separate issue.  Can we remove this for now?

File: dask/array/core.py (hunk 1 above)
Comment (itamarst) [reply dask/array/core.py]: This was not obvious to me, and this is... only place in the code where `_chunks` is set outside of `__new__`. Happy to remove it if you don't like it, though.

**Comment by itamarst** [comment]
I have removed hypothesis.

**Comment by martindurant** [comment]
There was one remaining request above, to remove the comments from the new code.

File: dask/array/core.py (hunk 1 above)
Comment (mrocklin) [reply dask/array/core.py]: Yeah, I'd prefer to remove it for now if that's ok

**Comment by martindurant** [comment]
Thank you!
Ready to merge when it goes green.

**Comment by mrocklin** [comment]
Thanks for the fix @itamarst and for the review @martindurant !

**Pull request merged**